### PR TITLE
Dynamic Haskell installation

### PR DESCRIPTION
**Pull request by meatballhat**
in preparation for GA of Haskell support on Trusty, where significantly fewer versions are preinstalled.

This PR, and the work done over in [travis-cookbooks](https://github.com/travis-ci/travis-cookbooks/blob/5fe919436e7db53b7b98a9a0871462fcb662a1ac/cookbooks/travis_build_environment/recipes/haskell.rb) is inspired by https://github.com/hvr/multi-ghc-travis.  If a `ghc` version is specified that is not already present, we will attempt to install it via `ppa:hvr/ghc`.

### regarding versions

Rather than depending on a preinstalled fuzzy version matching tool a la `nvm` or `rvm`, this PR includes `1.2.x` style version alias support like that currently in place for `language: go`.  The default `ghc` and `cabal` version values, which are `7.10.3` and `1.22`, respectively, are intended to match those of the [7.10.3 haskell platform release](https://www.haskell.org/platform/prior.html), which is the latest release available at the time of this writing, afaik.

/cc @hvr

### CODE REVIEW

File: lib/travis/build/config.rb [hunk 1]
```diff
@@ -33,15 +53,30 @@ def go_version_aliases_hash
...
-        cabal_default: ENV.fetch('TRAVIS_BUILD_CABAL_DEFAULT', '1.18'),
+        ghc_default: ENV.fetch('TRAVIS_BUILD_GHC_DEFAULT', '7.10.3'),
+        ghc_version_aliases: ENV.fetch(
+          'TRAVIS_BUILD_GHC_VERSION_ALIASES', (
+            latest_semver_aliases(
+              '6' => '6.12.3',
+              '7' => '7.10.3',
+              '8' => '8.0.2'
+            ).merge(
+              '7.8.x' => '7.8.4',
```
Comment (hvr) [inline lib/travis/build/config.rb]: why no 7.10.x and 8.0.x aliases?

Comment (meatballhat) [inline lib/travis/build/config.rb]: They should be there, but then I don't have specs in place, so shame on me :smile_cat:.  Are you OK with this approach?

Comment (hvr) [inline lib/travis/build/config.rb]: I assume this is about the values you canuse in https://docs.travis-ci.com/user/languages/haskell#Specifying-the-GHC-version ? If so, this looks good to me.

As for the rest of the script, I'm still digesting the code... :-)

**Comment by meatballhat** [comment]
Example of current behavior: https://staging.travis-ci.org/meatballhat/yolo-octo-adventure/jobs/530883

File: lib/travis/build/script/haskell.rb [hunk 2]
```diff
@@ -3,19 +3,38 @@ module Build
...
             template(
               'haskell.sh',
               default_ghc: DEFAULTS[:ghc],
+              default_cabal: DEFAULTS[:cabal],
               root: '/'
             )
           )
-          sh.export 'PATH', "#{path}:$PATH", assert: true
+          # Automatic installation of exact versions *only*.
+          if version =~ /^\d+\.\d+\.\d+$/ && cabal_version =~ /^\d+\.\d+$/
```
Comment (BanzaiMan) [inline lib/travis/build/script/haskell.rb]: Are these numbers reasonable? Do we ever have version numbers in GHC and cabal that don't match these?

Comment (meatballhat) [inline lib/travis/build/script/haskell.rb]: To my knowledge, this covers all GHC and cabal versions.

Comment (BanzaiMan) [inline lib/travis/build/script/haskell.rb]: All right, then.

**Comment by meatballhat** [comment]
@hvr Pending your approval 😺

**Comment by hvr** [comment]
@meatballhat I'd like to help by reviewing this, but I'm really lost in the ruby code... :-) otoh, once this is approved, can it be beta-tested somewhere? if so, I could try it out and review it "empirically"

**Comment by BanzaiMan** [comment]
@hvr If you can set up a test repo on https://staging.travis-ci.org, we can coordinate testing this.

File: lib/travis/build/config.rb (hunk 1 above)
Comment (phadej) [inline lib/travis/build/config.rb]: this map should have also `head` (and rest of code should cope with that). At least atm `cabal-install-head` is vital in some scenarious (not only new-build, but also solver improvements).

File: lib/travis/build/config.rb [hunk 3]
```diff
@@ -33,14 +53,30 @@ def go_version_aliases_hash
...
           ENV.fetch('TRAVIS_ENABLE_DEBUG_TOOLS', '')
         ),
         etc_hosts_pinning: ENV.fetch(
           'TRAVIS_BUILD_ETC_HOSTS_PINNING', ENV.fetch('ETC_HOSTS_PINNING', '')
         ),
-        ghc_default: ENV.fetch('TRAVIS_BUILD_GHC_DEFAULT', '7.8.4'),
+        ghc_default: ENV.fetch('TRAVIS_BUILD_GHC_DEFAULT', '7.10.3'),
+        ghc_version_aliases: ENV.fetch(
+          'TRAVIS_BUILD_GHC_VERSION_ALIASES', (
+            latest_semver_aliases(
```
Comment (phadej) [inline lib/travis/build/config.rb]: technically ghc versions aren't semver, if anything they are probably pvp. It's really hard to write non-trivial code which would work with say `ghc == 8.x.x`, as `8.2` will definitely break something.

I.e. I'd remove single version alias all-together, or actively discourage from using such.

**Comment by hvr** [comment]
@meatballhat so in order to try this out, I can just use http://staging.travis-ci.org the same way I use the normal travis-ci.org?

**Comment by meatballhat** [comment]
@hvr That's correct, although there are some gotchas.  For one, we recommend associating a new/dedicated repo with staging.travis-ci.org.  You *can* be safely signed into *both* travis-ci.org and staging.travis-ci.org.

File: lib/travis/build/config.rb [hunk 4]
```diff
@@ -8,13 +8,33 @@ class Config < Travis::Config
...
+            key, value = v.strip.split(':', 2)
+            next if key.nil? || value.nil?
+            aliases[key] = value
+          end
+        end
+      end
+
+      def self.latest_semver_aliases(major_full)
+        {}.tap do |aliases|
+          major_full.map do |major, full|
```
Comment (renee-travisci) [inline lib/travis/build/config.rb]: I would use each instead of map here since you aren't using the returned array.

**Comment by meatballhat** [comment]
@hvr @phadej bump ❤️

**Comment by phadej** [comment]
LGTM. I don't have time to try this unfortunately :(

**Comment by hvr** [comment]
@meatballhat I haven't forgotten about this, I'm just a bit busy; I'll try to get to try it as soon as I find a bit of time!

**Comment by emdantrim** [comment]
Hey folks,

We're going to go ahead and roll this out soon. We'll try and keep a close eye on things, so please let us know if anything goes awry. 🎉

**Comment by hvr** [comment]
@meatballhat @emdantrim I finally got to try this out; I'm using https://staging.travis-ci.org/hvr/travisnextgen for my experiments

I tried with a simple `.travis.yml` file

```yaml
language: haskell
ghc: 8.2.1
```

which failed right-away due to

```
Adding ppa:hvr/ghc.

Usage: apt-add-repository <sourceline>

apt-add-repository is a script for adding apt sources.list entries. 

It can be used to add any repository and also provides a shorthand 

syntax for adding a Launchpad PPA (Personal Package Archive)

repository.

<sourceline> - The apt repository source line to add. This is one of:

  a complete apt line in quotes, 

  a repo url and areas in quotes (areas defaults to 'main')

  a PPA shortcut.

  Examples:

    apt-add-repository 'deb http://myserver/path/to/repo stable myrepo'

    apt-add-repository 'http://myserver/path/to/repo myrepo'

    apt-add-repository 'https://packages.medibuntu.org free non-free'

    apt-add-repository http://extras.ubuntu.com/ubuntu 

    apt-add-repository ppa:user/repository

If --remove is given the tool will remove the given sourceline from your

sources.list

apt-add-repository: error: no such option: -q
```

```
$ export TRAVIS_HASKELL_VERSION=$(travis_ghc_find '8.2.1')
No such ghc version '8.2.1'.
```

**Comment by BanzaiMan** [comment]
This is already deployed to production. Please test it there now.

**Comment by hvr** [comment]
@BanzaiMan I assume production is this here: https://www.travis-ci.org/hvr/netrc/jobs/231649623#L120 ? If so, I'm afraid the same issue occurs there

**Comment by BanzaiMan** [comment]
@hvr Thanks for testing. Could you open a new issue at https://github.com/travis-ci/travis-ci/issues/new? Thanks.

**Comment by hvr** [comment]
@BanzaiMan done (filed as travis-ci/travis-ci#7752)

**Pull request merged**